### PR TITLE
Schemaless tap

### PR DESCRIPTION
**Pull request by JingLin0**

### CODE REVIEW

File: tap_g2crowd/g2crowd.py [hunk 1]
```diff
@@ -15,43 +15,66 @@ def __init__(self, config: Dict):
         self.config = config
         self.companies_endpoints = []
 
+    def do_sync(
+        self, tap_stream_id: str, stream_config: Dict, state: Optional[datetime] = None
+    ):
         stream = Stream(self.config["api_key"], self.companies_endpoints)
+        bookmark_key = stream_config.get("bookmark_key")
```
Comment (paddie) [inline tap_g2crowd/g2crowd.py:21]: we have the `bookmark_key` in the config?

File: tap_g2crowd/g2crowd.py [hunk 2]
```diff
@@ -15,43 +15,66 @@ def __init__(self, config: Dict):
...
-                        new_bookmark = replication_value
-                        if not prev_bookmark:
-                            prev_bookmark = new_bookmark
-
-                        if prev_bookmark < new_bookmark:
-                            state = self.__advance_bookmark(state, prev_bookmark)
-                            prev_bookmark = new_bookmark
-                self.companies_endpoints = stream.companies_endpoints
+                    if tap_stream_id == "remote_events_streams":
+                        self.companies_endpoints.append(
```
Comment (paddie) [inline tap_g2crowd/g2crowd.py:38]: maybe document why this is being done this way?

File: tap_g2crowd/g2crowd.py [hunk 3]
```diff
@@ -15,43 +15,66 @@ def __init__(self, config: Dict):
...
+                            stream.get_value(
+                                record, ["relationships", "company", "links", "related"]
+                            )
+                        )
+
+                    singer.write_record(tap_stream_id, record)
+                    counter.increment(1)
+                    if not replication_value:
+                        continue
```
Comment (paddie) [inline tap_g2crowd/g2crowd.py:47]: very good.

File: tap_g2crowd/g2crowd.py [hunk 4]
```diff
@@ -15,43 +15,66 @@ def __init__(self, config: Dict):
...
+                    if not prev_bookmark:
+                        prev_bookmark = new_bookmark
+
+                    if prev_bookmark < new_bookmark:
+                        state = self.__advance_bookmark(
+                            state=state,
+                            bookmark=prev_bookmark,
+                            bookmark_key=bookmark_key,
+                            tap_stream_id=tap_stream_id,
+                        )
```
Comment (paddie) [inline tap_g2crowd/g2crowd.py:58]: if we could `yield` the state every time it got updated. Then we don't need to do an explicit return at the end here. Perform the same yield in the exception, and then in the function that takes the state from the `do_sync` call would iterate over the returned states until we are done and use the final state for the next call..?

File: tap_g2crowd/stream.py [hunk 5]
```diff
@@ -12,39 +12,46 @@
...
 
     def get_companies(self):
+        # companies data is retrieved according to remote-event-streams
         if not self.companies_endpoints:
-            LOGGER.warn(
-                f"no remote_events_streams data to get companies data"
-            )
-            sys.exit(0)
+            LOGGER.warn(f"no remote_events_streams data to get companies data")
+            return None, None
```
Comment (paddie) [inline tap_g2crowd/stream.py:54]: yield

File: tap_g2crowd/stream.py [hunk 6]
```diff
@@ -99,6 +99,7 @@ def streams(self):
             requests.exceptions.HTTPError,
             ratelimit.exception.RateLimitException,
         ),
+        max_tries=5,
```
Comment (paddie) [inline tap_g2crowd/stream.py:95]: 👏

File: tap_g2crowd/stream.py [hunk 7]
```diff
@@ -102,14 +95,14 @@ def streams(self):
...
-    def call_api(self, url):
-
+    def call_api(self, url: str, params: Optional[Dict] = None):
         response = self.SESSION.get(
             url,
             headers={
                 "Accept": "application/vnd.api+json",
                 "Authorization": f"Token token={self.api_key}",
             },
+            params=params,
```
Comment (paddie) [inline tap_g2crowd/stream.py:105]: clean, nice

File: tap_g2crowd/g2crowd.py (hunk 1 above)
Comment (JingLin0) [reply tap_g2crowd/g2crowd.py:21]: it is `stream_config` not `config`. maybe I should name it differently.

**Pull request merged**